config/internal/configsource/configsource.go

```diff
@@ -0,0 +1,315 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configsource
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/spf13/viper"
+
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/config/internal/configsource/component"
+)
+
+const (
+	// ConfigSourcePrefix is used to identify a configuration source invocation.
+	ConfigSourcePrefix = "$"
+)
+
+// ApplyConfigSourcesParams holds the parameters for injecting data from the
+// given configuration sources into a configuration.
+type ApplyConfigSourcesParams struct {
+	// Set of configuration sources available to inject data into the configuration.
+	ConfigSources map[string]component.ConfigSource
+	// MaxRecursionDepth limits the maximum number of times that a configuration source
+	// can inject another into the config.
+	MaxRecursionDepth uint
+}
+
+// ApplyConfigSources takes a viper object with the configuration to which the
+// the configuration sources will be applied and the resulting configuration is
+// returned as a separated object.
+func ApplyConfigSources(ctx context.Context, v *viper.Viper, params ApplyConfigSourcesParams) (*viper.Viper, error) {
+
+	// Notify all sources that we are about to start.
+	for cfgSrcName, cfgSrc := range params.ConfigSources {
+		if err := cfgSrc.BeginSession(ctx); err != nil {
+			return nil, &cfgSrcError{
+				msg:  fmt.Sprintf("config source %s begin session error: %v", cfgSrcName, err),
+				code: errCfgSrcBeginSession,
+			}
+		}
+
+		// The scope of usage of the cfgSrc is the whole function so it is fine to defer
+		// the EndSession call to the function exit.
+		defer cfgSrc.EndSession(ctx)
+	}
+
+	srcCfg := v
+	var dstCfg *viper.Viper
+
+	var err error
+	done := false
+	for i := -1; i <= int(params.MaxRecursionDepth) && !done; i++ {
```

Comment:
Am I reading this correctly that you are doing `MaxRecursionDepth + 2` iterations?

Reply:
yeah, this is confusing. The code does at minimum two passes when there are config sources: one to resolve the original config sources and a second to check that no more config sources need to be applied. In a sense, the recursion starts with the first pass introduces more config source applications. I will look at a way to simplify this.

Reply:
Resolving since this part was removed from the PR.